### PR TITLE
Refactor Iceberg tests and introduce in-memory catalog

### PR DESCRIPTION
**Pull request by jackye1995**
This PR refactors all tests into `hivemode` and `nativemode` to have more test coverage for the native mode connector before I introduce more features to the connector. 

An `InMemoryCatalog` is also introduced for testing instead of using `HadoopCatalog`. `HadoopCatalog` is generally discouraged for production use by the Iceberg community, and have a lot of behaviors that are different from other catalog implementations. For example, it cannot set namespace property, rename table, do not have true transaction in local mode, etc. By switching to the in-memory catalog, we are able to test the native connector with much more confidence that the same connector can work for other catalog implementations that requires service connection and cannot be easily tested.

We will eventually have the in-memory catalog in the Iceberg library, but I am directly offering an implementation in this PR in the Presto codebase to speed up the delivery.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by jackye1995** [comment]
@ChunxuTang @beinan

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/CatalogType.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
 {
     HADOOP(HadoopCatalog.class.getName()),
     HIVE(HiveCatalog.class.getName()),
+    MEMORY(InMemoryCatalog.class.getName())
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/CatalogType.java:23]: Has the `InMemoryCatalog` merged into the Iceberg? If not, do you have a possible target ETA in your mind? In the future, I think we can directly import it from the Iceberg and avoid implementing the detailed catalog on the Presto side.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/hivemode/TestHiveIcebergOrcMetricsCollection.java [hunk 2]
```diff
@@ -11,9 +11,10 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.facebook.presto.iceberg;
+package com.facebook.presto.iceberg.hivemode;
```
Comment (ChunxuTang) [inline presto-iceberg/src/test/java/com/facebook/presto/iceberg/hivemode/TestHiveIcebergOrcMetricsCollection.java:14]: I noticed that for the native mode, there's no corresponding `TestNativeIcebergOrcMetricsCollection` file. Does the native mode support it?

**Review by ChunxuTang — COMMENTED** [review]
@jackye1995 Thanks for your nice contribution!
Just left some minor comments.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.iceberg.CatalogProperties.WAREHOUSE_LOCATION;
+import static org.apache.iceberg.CatalogUtil.dropTableData;
+import static org.apache.iceberg.CatalogUtil.loadFileIO;
+
+/**
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:47]: do we need this comment? Or we could remove it

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+ * An in memory catalog implementation to test Iceberg native mode
+ */
+public class InMemoryCatalog
+        extends BaseMetastoreCatalog
+        implements SupportsNamespaces, Catalog
+{
+    private final ConcurrentHashMap<Namespace, Map<String, String>> namespaces = new ConcurrentHashMap<>();
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:54]: ConcurrentMap<Namespace, ...> namespaces = new ConcurrentHashMap<>();

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    private final ConcurrentHashMap<TableIdentifier, String> tables = new ConcurrentHashMap<>();
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:55]: same for ConcurrentMap

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    @Override
+    public String name()
+    {
+        return "memory";
+    }
+
+    @Override
+    public void initialize(String name, Map<String, String> properties)
+    {
+        this.warehouseLocation = requireNonNull(properties.get(WAREHOUSE_LOCATION), "warehouseLocation is null");
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:69]: remove this?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        String fileIOImpl = properties.get(CatalogProperties.FILE_IO_IMPL);
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:70]: static import CatalogProperties.FILE_IO_IMPL
Presto variables not using abbriviation:
s/fileIOImpl/fileIOImplementation/g

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+        if (fileIOImpl == null) {
+            this.fileIO = new HadoopFileIO(new Configuration());
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:72]: remove this

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+            fileIO.initialize(properties);
+        }
+        else {
+            this.fileIO = loadFileIO(fileIOImpl, properties, new Configuration());
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:76]: remove this

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+    @Override
+    public boolean dropNamespace(Namespace namespace)
+            throws NamespaceNotEmptyException
+    {
+        validateNamespaceExists(namespace);
+        if (!listTables(namespace).isEmpty()) {
+            throw new NamespaceNotEmptyException("Namespace %s contains tables: %s", namespace, tables);
+        }
+        namespaces.remove(namespace);
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:116]: do we need to check permission when removing namespace?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+    protected TableOperations newTableOps(TableIdentifier tableIdentifier)
+    {
+        return new InMemoryTableOperations(fileIO, tableIdentifier);
+    }
+
+    @Override
+    protected String defaultWarehouseLocation(TableIdentifier tableIdentifier)
+    {
+        String tableName = tableIdentifier.name();
+        StringBuilder sb = new StringBuilder();
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:166]: s/sb/builder/g

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+
+        sb.append(warehouseLocation).append('/');
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:168]: shall we declare '/' as constant?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+        }
+        sb.append(tableName);
+        return sb.toString();
+    }
+
+    @Override
+    public boolean dropTable(TableIdentifier tableIdentifier, boolean purge)
+    {
+        validateTableExists(tableIdentifier);
+        TableOperations ops = newTableOps(tableIdentifier);
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:180]: s/ops/operations/g

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+        TableMetadata lastMetadata = ops.current();
+        tables.remove(tableIdentifier);
+        if (purge) {
+            dropTableData(fileIO, lastMetadata);
+        }
+        return true;
+    }
+
+    @Override
+    public void renameTable(TableIdentifier from, TableIdentifier to)
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:190]: s/to/tableIdentifier/g

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+                }
+            }
+            refreshFromMetadataLocation(metadataLocation);
+        }
+
+        @Override
+        protected void doCommit(TableMetadata base, TableMetadata metadata)
+        {
+            String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+            synchronized (tables) {
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:257]: tables is ConcurrentMap, why we need synchronized?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+                String inMemoryLocation = tables.getOrDefault(tableIdentifier, null);
+                String baseLocation = base != null ? base.metadataFileLocation() : null;
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/InMemoryCatalog.java:259]: could we use Optional<String> instead of String and null?

**Review by zhenxiao — CHANGES_REQUESTED** [review]
nice work, @jackye1995 
left some comments. Could you please resolve @ChunxuTang 's comments and resolve conflicts?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/CatalogType.java (hunk 1 above)
Comment (nastra) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/CatalogType.java:23]: just fyi, the `InMemoryCatalog` hasn't been merged yet (see https://github.com/apache/iceberg/pull/4054)

**Comment by hantangwangd** [comment]
Closing this PR for there is no need to implement a `InMemoryCatalog` in presto side. Iceberg has implemented and merged its `InMemoryCatalog` through https://github.com/apache/iceberg/pull/6927

**Pull request closed without merging**